prime-router/src/main/kotlin/azure/HistoryFunctions.kt

```diff
@@ -290,11 +290,27 @@ open class BaseHistoryFunction {
                 val mimeType = Report.Format.safeValueOf(header.reportFile.bodyFormat).mimeType
 
                 val fileReturn = FileReturn( String(header.content), filename, mimeType);
-                return request
+                response = request
                     .createResponseBuilder(HttpStatus.OK)
                     .header("Content-Type", "application/json")
                     .body( fileReturn )
                     .build()
+
+                val actionHistory = ActionHistory(TaskAction.download, context)
+                actionHistory.trackActionRequestResponse(request, response)
+                // Give the external report_file a new UUID, so we can track its history distinct from the
+                // internal blob.   This is going to be very confusing.
+                val externalReportId = UUID.randomUUID()
+                actionHistory.trackDownloadedReport(
+                    header,
+                    filename,
+                    externalReportId,
+                    authClaims.userName,
+                )
+                actionHistory.trackItemLineages(Report.createItemLineagesFromDb(header, externalReportId))
+                WorkflowEngine().recordAction(actionHistory)
+
+                return response
```

Comment:
nitpick: remove the return, and let it drop into line 322 which returns.

Reply:
Yes, this could be more kotlinesque, but is ok for now.   Perhaps normal Kotlin style would be:

return try {
   ...
}

so that you don't even need a response variable.